vermouth/gmx/topology.py

```diff
@@ -0,0 +1,216 @@
+
+def _group_by_conditionals(interactions):
+    interactions_group_sorted = sorted(interactions,
+        key=_interaction_sorting_key
+        )
+    interaction_grouped = itertools.groupby(
+        interactions_group_sorted,
+        key=_interaction_sorting_key
+        )
+    return interaction_grouped
+
+def convert_sigma_epsilon(sigma, epsilon):
+    """
+    Convert the LJ potential from sigma epsilon
+    form to C6 C12 form.
+    """
+    C6 = 4*sigma*epsilon**6
+    C12 = 4*sigma*epsilon**12
+    return C6, C12
+
+def write_atomtypes(system, itp_path, C6C12=False):
+    """
+    Writes the [atomtypes] directive to file.
+    All atomtypes are defined in system.gmx_topology_params.
+    Masses and further information are taken from the molecule
+    directly.
+    """
+    conditional_keys = {True: '#ifdef', False: '#ifndef'}
+    with deferred_open(itp_path, "w") as itp_file:
+        itp_file.write("[ atomtypes ]\n")
+        grouped_types = _group_by_conditionals(system.gmx_topology_params['atomtypes'])
+        for (conditional, group), interactions_in_group in grouped_types:
+            if conditional:
+                conditional_key = conditional_keys[conditional[1]]
+                itp_file.write('{} {}\n'.format(conditional_key, conditional[0]))
+            if group:
+                itp_file.write('; {}\n'.format(group))
+
+            for atomtype in interactions_in_group:
+                atype = atomtype.molecule.nodes[atomtype.node]['atype']
+                charge = atomtype.molecule.nodes[atomtype.node]['charge']
+                mass = atomtype.molecule.nodes[atomtype.node]['mass']
+
+                if C6C12:
+                    nb1, nb2 = convert_sigma_epsilon(atomtype.sigma, atomtype.epsilon)
+                else:
+                    nb1, nb2 = atomtype.sigma, atomtype.epsilon
+
+                if 'comment' in atomtype.meta:
+                    comments = ";" + " ".join(atomtype.meta['comment'])
+                else:
+                    comments = ""
+                itp_file.write(f"{atype} {mass} {charge} A {nb1:3.8F} {nb2:3.8F} {comments}\n")
+
+def write_nonbond_params(system, itp_path, C6C12=False):
+    """
+    Writes the [nonbond_params] directive to file.
+    All atomtypes are defined in system.gmx_topology_params.
+    Masses and further information are taken from the molecule
+    directly.
+    """
+    conditional_keys = {True: '#ifdef', False: '#ifndef'}
+    with deferred_open(itp_path, "w") as itp_file:
+        itp_file.write("[ nonbond_params ]\n")
+        grouped_types = _group_by_conditionals(system.gmx_topology_params['nonbond_params'])
+        for (conditional, group), interactions_in_group in grouped_types:
+            if conditional:
+                conditional_key = conditional_keys[conditional[1]]
+                itp_file.write('{} {}\n'.format(conditional_key, conditional[0]))
+            if group:
+                itp_file.write('; {}\n'.format(group))
```

Comment:
Is it worth the headache to deduplicate this?

Reply:
probably but this is kinda cosmetic so I'd rather not now